platform/vita/vita_osal.c

```diff
@@ -236,15 +236,18 @@ pte_osResult pte_osThreadWaitForEnd(pte_osThreadHandle threadHandle)
 {
 	int status = 0;
 	pspThreadData *pThreadData = *(pspThreadData **)vitasdk_get_pthread_data(0);
-	SceUID evid = pThreadData->evid;
```

Comment:
can you explain why we need to allow calling it for non-pthread thread?
how about just
```
if (!pThreadData) {
  return PTE_OS_GENERAL_FAILURE;
}
```
this?

Reply:
The situation causing me trouble goes something like this:
```mermaid
sequenceDiagram
    participant main thread
    participant thread 1
    participant thread 2
    main thread->>thread 1: sceKernelCreateThread
    thread 1->>thread 2: pthread_create
    thread 1->>thread 2: pthread_join
    main thread->>thread 1: sceKernelWaitThreadEnd
```
Which IMO is perfectly OK. `Thread 1` crashes when calling `pthread_join` on `Thread 2`.
Note that the crash does not happen when the main thread uses pthread directly because https://github.com/vitasdk/pthread-embedded/blob/3ad4d887b84e32a226d9dffa019e4c7b229da785/platform/vita/vita_osal.c#L651 creates a dummy `pspThreadData` structure for the main thread on program start.

> how about just
> 
> ```
> if (!pThreadData) {
>   return PTE_OS_GENERAL_FAILURE;
> }
> ```
> 
> this?

I would prefer a fix that goes in the direction of making things work rather than the opposite.

Reply:
@d3m3vilurr do you have any updates regarding this PR?

Reply:
I understand your case for this. but i'm worried it has side effect or not, i still don't know that mixing sce thread and posix thread is normal or not.
so, I think it need more review from other maintainers.
@frangarcj can you review this?

Reply:
why do you create the sce thread first?

Reply:
I don't create any threads in the main program. This is all libraries and ports creating threads. As a user, I don't care how the threading is implemented in the various libraries. I want each library to be able to manage its threads even if some libs use the native threads directly and some use them through pthread.

Reply:
i'm still not sure that this patch can fix the problem.

@bgK can you share build options of your program? and if application try overriding weakref functions, please also share information of it

Reply:
Sure, build options:
`CFALGS=-g -Wl,-q -O2 -fno-lto -Wl,--whole-archive pthread -Wl,--no-whole-archive [other static libraries go here]`

I am not overriding library functions.

Reply:
if other staticlibs & your application don't override weakref function, `main` should be called `pthread_init`.

```
// newlib/libc/sys/vita/threading.c
void _init_vita_reent(void)
{
  ...
}

// newlib/libc/sys/vita/crt0.c
void _init_void_newlib(void)
{
  _init_vita_reent();
}

...

void _start(unsigned int args, void *argp)
{
  ...
  _init_vita_newlib();
  ...
}
```

this chain could be initialize default thread datas.
did you try to call pthread_join with non-pthread data?
and also if this issue heppend with external lib & the lib is opensource, can you share the name of the external lib

Reply:
> this chain could be initialize default thread datas.

Yes, the global Newlib thread data is properly initialized. What's `NULL` is the per-thread pthread data for the non-pthread thread (seems normal to me).

>  and also if this issue heppend with external lib & the lib is opensource, can you share the name of the external lib

The threading issue is between SDL2 (Vita native) and an Android shared object making use of pthread.

Reply:
aha..... so normal vita version sdl2 uses vita native thread(maybe it for a speed benefit or pthread wasn't default of the toolchain at that time). but android sdl2 uses pthread & maybe somewhere also call the pthread.
these different made a problem.

couple weeks ago, @Rinnegatamante shared to me as pretty similar cases in enet porting. if vita uses modified enet, the structure size would be different between vita and android.

imo, best option is to rebuild sdl with modified cmakefile & src/thread/pthread (and btw needs more test of sdl part)
if you really want to modify the pthread, just build your own version and install into vitasdk/arm-vita-eabi/lib/pthread (or manually linking the lib file)
and another option is making more fake layers for pthread & sdl
for example simply if you only have a problem when call the `pte_osThreadWaitForEnd`
you can make the `pthread_join_fake` & `pthread_detach_fake`

```
// actual struct can be found in implement.h
struct fake_pte_thread_t {
  SceUID thread_id;
};

int pthread_join_fake(pthread_t thread, void **value_ptr) {
  struct fake_pte_thread_t *tp = (struct fake_pte_thread_t *)thread;
  // need mutex with `pte_thread_reuse_lock`
  ...

  void *p_thread_data = *(void**)vitasdk_get_pthread_data(0);
  if (p_thread_data) {
    return pthread_join(thread, value_ptr);
  }
  while (1) {
    // ...
    int res = sceKernelWaitThreadEndCB(tp->thread_id, &status, &timeout);
    ...
  }
  ...
  
  return 0;
}

static DynLibFunction dynlib_functions[] = {
...
  {"pthread_join", (uintptr_t)&pthread_join_fake},
...
```

actually, i'm still thinking that it's not a bug of the pthread-embedded part in normal environment.

@Rinnegatamante do you have any additional tips to share with @bgK ?

Reply:
I mean, sure there are workarounds. There are no mismatching structure sizes here, it's a simple null dereference in the pthread-embedded implementation.

What's wrong with the fix proposed in this PR?

Reply:
@bgK sorry, i still think that this is not a problem of the pthread-embedded.
the problem is someone passing non-pthread value into here.

in this case, i think you should need more wrapping of sdl2 functions. and then probably gone the problem.

the bug in this pr is access nullish object in logic because some code passing the wrong value into here.
it means, a problem can be confirmed. plus solutions to prevent the crash are

1. check the type of value to prevent passing the invalid value
2. ignore them if the value is wrong

if you make/embed your thread handing code into your project, you can choice 2. but i don't agree, it should be 1.

(first option is same as my first comment
it's same as my first comment 
```
how about just

> ```
> if (!pThreadData) {
>   return PTE_OS_GENERAL_FAILURE;
> }
> ```

this?
```
)

idk i'm right or not, your program might make the thread using `SDL_SYS_CreateThread` (or `SDL_CreateThread`) and you also mapped this function,
but you didn't map the `SDL_SYS_WaitThread` (or didnt map `SDL_WaitThread`) and you just wrap the `pthread_join` instead them.
(or compiler optimizer might be removed SDL_SYS_WaitThread... but idk)

Reply:
This is a minimal reproduction case for the bug. No SDL, no Android, just pthreads:
```c
#include <psp2/kernel/threadmgr.h>
#include <psp2/kernel/processmgr.h>
#include <stdio.h>
#include <pthread.h>

static void *thread_2_proc(void *arg) {
	sceKernelDelayThread(3*1000000);
    return NULL;
}

static int thread_1_proc(SceSize args, void *argp) {
	pthread_t thread_2;
	pthread_create(&thread_2, NULL, &thread_2_proc, NULL);
	pthread_join(thread_2, NULL);
	
	return 0;
}

int main(int argc, char *argv[]) {
	printf("Start!\n");
	
	int thread_1_id = sceKernelCreateThread("thread_1", &thread_1_proc, 0x10000100, 0x10000, 0, 0, NULL);
	sceKernelStartThread(thread_1_id, 0, 0);
	sceKernelWaitThreadEnd(thread_1_id, NULL, NULL);
	
	printf("End!\n");
	return 0;
}
```

When running this program without this PR, the following crash occurs:
```
Thread 1: thread_1 (Running) - Data abort exception at PC 0x8104bf1c reading memory at 0x00000008
  0x81002f1c   (pthread_test): in pte_osThreadWaitForEnd+0x1c at pthread-embedded/platform/vita/vita_osal.c:239
  0x810016df   (pthread_test): in pthread_join+0x5b at pthread-embedded/pthread_join.c:133
  0x810001e3   (pthread_test): in thread_1_proc+0x1b at pthread_test/src/main.c:15
  0xe000f039   (SceLibKernel): ??
```

Program with build script and other required files:
[pthread_test.zip](https://github.com/vitasdk/pthread-embedded/files/9267712/pthread_test.zip)

Is this sample program incorrect?

Reply:
Crashed because TLS key 0x89 was not set. There is also vitasdk code as failback, but they all fail.

Reply:
> Crashed because TLS key 0x89 was not set. There is also vitasdk code as failback, but they all fail.

TLS key 0x89 seems set to me. The crash is because `pthread_data_ext` is `NULL` in the struct `reent_for_thread`:
https://github.com/vitasdk/newlib/blob/4b63470c84aee0e8bc543e6b764c874caa0b7ac3/newlib/libc/sys/vita/threading.c#L19
(As the thread `thread_1` is not a pthread thread)

Reply:
maybe
```
pspThreadData *pThreadData = *(pspThreadData **)vitasdk_get_pthread_data(0);
```
of the `pte_osThreadWaitForEnd` to be
```
pspThreadData *pThreadData = *(pspThreadData **)vitasdk_get_pthread_data(threadHandle);
```

Reply:
```c
pspThreadData *pThreadData = *(pspThreadData **)vitasdk_get_pthread_data(0);
````
is correct. As explained in this comment, the goal of the event flag polling loop is to allow the `pthread_join` operation to be cancelled:
https://github.com/vitasdk/pthread-embedded/blob/master/platform/vita/vita_osal.c#L232

This implies three threads:
* Thread A doing some work
* Thread B waiting for Thread A to complete its work using `pthread_join(thread_a)`
* Thread C cancelling the wait of Thread B using `pthread_cancel(thread_b)`

This is why the event flag of the thread calling `pthread_join` is used (Thread B), and not that of the thread being waited upon (Thread A).

The point of this PR is that Thread B (Thread 1 in the minimal repro case) is not a pthread thread. Thus, `pthread_cancel` won't be called on it. Which means there is no need to check the event flag. Hence the proposed fix.

Reply:
hm... i also thought that(but idk, i might think wrongly.)

i thought
1. A made B, B made C, B wait C, and A wait B (same as your sample)
2. A made B, B made C, B wait C, and A canceled B's request (or whatever another thread). then A wait B (it could be same as your comment)

a thread cancel is related implementation of the `pte_osThreadCancel` and it register the eventid
```
pte_osResult pte_osThreadCancel(pte_osThreadHandle threadHandle)
{
	pspThreadData *pThreadData = *(pspThreadData **)vitasdk_get_pthread_data(threadHandle);

	int res = sceKernelSetEventFlag(pThreadData->evid, PTHREAD_EVID_CANCEL);

	if (res < 0)
		return PTE_OS_GENERAL_FAILURE;

	return PTE_OS_OK;
}
```
but this cancel request has registered to target threadid (not the base TLS).
so i think it's another bug that this cancel wouldn't work..

currently only two functions were used base TLS that they are `pte_osSemaphoreCancellablePend` and `pte_osThreadWaitForEnd`

but original code of them for psp was used current thread id
```
pThreadData = getThreadData(sceKernelGetThreadId());
```

Reply:
Wait, there's no such thing as "base TLS". The `0` in `vitasdk_get_pthread_data(0)` means "for the current thread", as seen here:
https://github.com/vitasdk/newlib/blob/4b63470c84aee0e8bc543e6b764c874caa0b7ac3/newlib/libc/sys/vita/threading.c#L154

The Vita implementation is equivalent to the PSP one.

> (and i think semaphore for vita should need more implements)

Yes, `pte_osSemaphoreCancellablePend` needs the same fix as I did in `pte_osThreadWaitForEnd`. I can add the change if you want.

Reply:
> Wait, there's no such thing as "base TLS". The `0` in `vitasdk_get_pthread_data(0)` means "for the current thread", as seen here: https://github.com/vitasdk/newlib/blob/4b63470c84aee0e8bc543e6b764c874caa0b7ac3/newlib/libc/sys/vita/threading.c#L154
> 

it doesn't mean current thread.
it means return the TLS data
that because probably `*on_tls` exists after `_init_vita_reent`
```
*(struct _reent **)(TLS_REENT_PTR) = &reent_list[0].reent;
```

> The Vita implementation is equivalent to the PSP one.
> 
> > (and i think semaphore for vita should need more implements)
> 
> Yes, `pte_osSemaphoreCancellablePend` needs the same fix as I did in `pte_osThreadWaitForEnd`. I can add the change if you want.

it should be an another pr not in here :)

Reply:
> it doesn't mean current thread.
> it means return the TLS data

```c
	if (thid == 0)
		on_tls = TLS_REENT_PTR;
	else
		on_tls = TLS_REENT_THID_PTR(thid);	
```
```c
#define TLS_REENT_THID_PTR(thid)	sceKernelGetThreadTLSAddr(thid, 0x89)
#define TLS_REENT_PTR				sceKernelGetTLSAddr(0x89)
```
The documentation for `sceKernelGetTLSAddr` says:
> sceKernelGetTLSAddr get pointer to TLS key area for current thread

Reply:
oh.... yeah...

`pspThreadData *pThreadData = *(pspThreadData **)vitasdk_get_pthread_data(threadHandle);`
is not a same as
`pspThreadData *pThreadData = *(pspThreadData **)vitasdk_get_pthread_data(sceKernelGetThreadId());`

i have to figure out why thread 1's data was broken.. :'(

Reply:
> i have to figure out why thread 1's data was broken.. :'(

`thread_1`'s data was not broken, it never had a `pspThreadData` instance to begin with because it is not a pthread thread. (`pte_osThreadCreate` allocates the structure when `pthread_create` is called)

Reply:
oh shit. right. that's what my misunderstanding.
i'll think that we can inject the value to parent when calling `pthread_create` time
(i means, even if thread_1 is not a pthread thread, it should have tls information after calling `pthread_create` like `main` function)
it can't inject data, i'll merge this.
plus i'll change the title and description of the pr

Reply:
> i'll think that we can inject the value to parent when calling `pthread_create` time
> (i means, even if thread_1 is not a pthread thread, it should have tls information after calling `pthread_create` like `main` function)

There's also the case where the thread calling `pthread_join` is not the thread that called `pthread_create`.

Reply:
really do we need to handle that case too? :thinking: 

and how about 11a4638? imo it can also resolve this issue.

Reply:
> really do we need to handle that case too? thinking

I don't need it for what I'm doing at the moment. I would handle it though, so people don't encounter the issue in the future.

> and how about https://github.com/vitasdk/pthread-embedded/commit/11a46380070a3f3e28dc112129bb4e6114006162? imo it can also resolve this issue.

It solves my issue.

Reply:
> > really do we need to handle that case too? thinking
> 
> I don't need it for what I'm doing at the moment. I would handle it though, so people don't encounter the issue in the future.
> 
yes. we have to add more errors in similar cases than crash fault.

> > and how about [11a4638](https://github.com/vitasdk/pthread-embedded/commit/11a46380070a3f3e28dc112129bb4e6114006162)? imo it can also resolve this issue.
> 
> It solves my issue.

so can I replace this pr to it? if you don't have any argue, i'll make the new pr for reviewing.

Reply:
> yes. we have to add more errors in similar cases than crash fault.

I mean I would make sure it works, not make it fail.

> so can I replace this pr to it? if you don't have any argue, i'll make the new pr for reviewing.

Who am I to ague, this is not my project.